src/treeffuser/_base_tabular_diffusion.py

```diff
@@ -97,41 +99,137 @@ def get_new_score_model(self) -> ScoreModel:
+                            f"{set(self._x_dataframe_columns) - set(X.columns)}"
+                        )
+                    # reorder the columns of the DataFrame to match the order of the training data
+                    X = X[self._x_dataframe_columns]
+                X = X.values
+
+            elif isinstance(X, np.ndarray):
+                if cat_idx is not None:
+                    for idx in cat_idx:
+                        if idx < 0 or idx >= X.shape[1]:
+                            raise ValueError(
+                                f"Invalid indices in `cat_idx`: {idx} is not in "
+                                f"[0, {X.shape[1]}-1] (the shape of X)."
+                            )
+                else:
+                    cat_idx = []
+
+                if reset:
+                    self._x_cat_idx = cat_idx
+            else:
+                raise TypeError(
+                    "Input data `X` must be a numpy array or a pandas DataFrame."
+                    f"Reveived {type(X).__name}."
+                )
+
             X = _check_array(X)
 
-        if validate_y and y is not None:
+        if validate_y:
+            if y is None:
+                raise ValueError("Target data `y` cannot be None.")
+            if isinstance(y, pd.Series):
+                y = y.values
+            elif not isinstance(y, np.ndarray):
+                raise TypeError(
+                    "Target data `y` must be a numpy array or a pandas Series."
+                    f"Received {type(y).__name}."
+                )
+
+            if reset:
+                # store the original number of dimensions of the response
+                self._y_original_ndim = y.ndim
+                self._y_dim = y.shape[1] if y.ndim > 1 else 1
+
             y = _check_array(y)
 
-        return X, y
+        return X, y, self._x_cat_idx
 
     def fit(
         self,
-        X: Float[ndarray, "batch x_dim"],
-        y: Float[ndarray, "batch y_dim"],
+        X: Union[Float[ndarray, "batch x_dim"], pd.DataFrame],
+        y: Union[Float[ndarray, "batch y_dim"], pd.Series],
         cat_idx: Optional[List[int]] = None,
     ):
         """
         Fit the conditional diffusion model to the tabular data (X, y).
 
         Parameters
         ----------
-        X : np.ndarray
+        X : np.ndarray or pd.DataFrame
             Input data with shape (batch, x_dim).
-        y : np.ndarray
+        y : np.ndarray or pd.Series
             Target data with shape (batch, y_dim).
         cat_idx : List[int], optional
-            List of indices of categorical features in X. Default is None.
+            If X is a np.ndarray, list of indices of categorical features in X.
+            If X is a DataFrame, setting `cat_idx` will raise an error. Instead, ensure that the
+            categorical columns have dtype `category`, and they will be automatically detected as
```

Comment:
ntp. The example you provided in the value error above could be nice to add here

Reply:
good idea, done.